[SPARK-10620][WIP] Migrate TaskMetrics to accumulators

andrewor14 opened the pull request:
There exist two mechanisms to pass metrics from executors to drivers: accumulators and `TaskMetrics`. Currently we send both things to the driver using two separate code paths. This is an unnecessary maintenance burden and makes the code more difficult to follow.

**This patch proposes that we send only accumulator updates to the driver.** Additionally, it reimplements `TaskMetrics` using accumulators such that the new `TaskMetrics` serves mainly as a syntactic sugar to increment and access the values of the underlying accumulators. It migrates the rest of the metrics to adopt the code path already used by the existing `PEAK_EXECUTION_MEMORY`.

While an effort has been made to preserve as much of the public API as possible, there were a few known breaking `@DeveloperApi` changes that would be very awkward to maintain. These are:
- `TaskMetrics#hostname` field was removed; event log was the only consumer
- `ExceptionFailure#taskMetrics` field was replaced with `accumUpdates`
- `SparkListenerExecutorMetricsUpdate#taskMetrics` field was replaced with `accumUpdates`
- `AccumulableInfo#update` changed from `Option[String]` to `Option[Any]`
- `AccumulableInfo#value` changed from `String` to `Option[Any]`

The following event log elements are changed:
- `AccumulableInfo` "Update" field changed from `Option[String]` to `Option[Any]`
- `AccumulableInfo` "Value" field changed from `String` to `Option[Any]`

This is WIP because I would like to add tests for some of the intricate cases that I ran into while implementing this.

---

I broke this down into several smaller PRs:
- #10810 - move classes to their own files for readability
- #10811 - rename 3 fields in `ShuffleWriteMetrics`
- #10815 - refactor TaskMetrics to make it not use public var's
- #10835 - implement TaskMetrics using accumulators
- #10857 - send only accumulator updates to driver, not TaskMetrics

Commit:
Remove unused methods or replace them

There are a bunch of decrement X methods that were not used.
Also, there are a few set X methods that could have easily just
been increment X. The latter change is more in line with
accumulators.

Commit:
Implement initial framework to migrate metrics to accums

This commit uses the existing PEAK_EXECUTION_MEMORY mechanism
to bring a few other fields in TaskMetrics to use accumulators.

Commit:
Migrate a few more easy metrics

Commit:
ShuffleReadMetrics + namespacing accumulators

This commit ports ShuffleReadMetrics to use accumulators,
preserving as much of the existing semantics as possible.
It also introduces a nicer way to organize all the internal
accumulators by namespacing them.

Commit:
General code cleanup

Commit:
ShuffleWriteMetrics

Commit:
OutputMetrics

Commit:
InputMetrics

This commit was a little tricky because it ripped the bytes read
callback from TaskMetrics and related classes. It does change
behavior in the sense that now we periodically update the number
of bytes read (every 1000 records) instead of doing it every time
we send an executor heartbeat. The advantage here is code
simplicity.

Commit:
Fix JsonProtocol + JsonProtocolSuite

Commit:
Fix tests where TaskMetrics had no accumulators

Tests are still failing as of this commit. E.g. SortShuffleSuite.

Commit:
Rename a few shuffle write metrics for consistency

Commit:
Fix metrics in local mode (tests)

Tests were previously failing because we end up double counting
metrics in local mode. This is because each TaskContext shares
the same list of accumulators, so they end up updating the metrics
on top of each other. The fix is to ensure TaskContext clears any
existing values on the accumulators before passing them on.

Commit:
Fix "harmless" exception in peak memory tests

The exception was harmless because it didn't actually fail the
test. However, the test harness was actually badly written. We
used to always assume that the first job will have an ID of 0,
but there could very well be other tests sharing the same
SparkContext. This is now fixed and we no longer see the exception.

As of this commit, all known test failures have been fixed.
I'm sure there will be more...

Commit:
Simplify internal accumulator update mechanism

Instead of passing in a callback, we can just return the
accumulator values directly, which we have. "We" here refers to
TaskMetrics.

Commit:
Fix tests

Commit:
Clean up

This commit addresses outstanding TODO's and makes the deprecated
APIs DeveloperApi instead. This allows us to deal with how to
do the deprecation properly later. This commit also reverts a few
unnecessary changes to reduce the size of the diff.

Commit:
Merge branch 'master' of github.com:apache/spark into task-metrics-to-accums

Conflicts:
	core/src/main/scala/org/apache/spark/TaskContextImpl.scala
	sql/core/src/test/scala/org/apache/spark/sql/execution/ui/SQLListenerSuite.scala

Commit:
Do not send TaskMetrics from executors to driver

Instead, send only accumulator updates. As of this commit
TaskMetrics is only used as a syntactic sugar on the executor
side to modify accumulator values by names. Now we no longer
send the same thing in two different codepaths.

Now that we never send TaskMetrics from executors to the driver,
we also never send accumulators that way. Then we can revert some
of the accumulator changes.

Commit:
Restore accumulator serialization semantics

In the previous commit, we made accumulator communication one-way
again, which is the same as before this patch, so we restored all
the semantics involved in serializing accumulators as before.

Note: tests are still failing because of a duplicate accumulator
name in some SQL things. Run `DataFrameCallbackSuite` for more
detail.

Commit:
Fix semantics of accumulators when tasks fail

Currently we still get values for tasks that fail. We should
keep this semantics in the new accumulator updates as well.

Commit:
Fix a few more tests

There are a few places where we passed in empty internal
accumulators to TaskContextImpl, so the TaskMetrics creation
would fail. These are now fixed.

Commit:
Fix SQL UI

Before this commit the SQL UI would not display any accumulators.
This is because it is powered by the SQLListener, which reads
accumulators from TaskMetrics. However, we did not update the
accumulator values before posting the TaskMetrics, so the UI
never saw the updates from the tasks.

This commit also fixes a few related test failures.

Commit:
Clean up: lift odd unique name requirement

Now internal accumulators no longer need to have unique names.
This was an unnecessary hack for the SQL accumulators that can
be reverted through some clean ups.

Commit:
Move smaller metrics classes to their own files

for readability.

Commit:
Fix SQLQuerySuite

A few bugs:

(1) In Executor.scala, we updated TaskMetrics after collecting the
accumulator values. We should do it the other order.

(2) The test utility method of verifying whether peak execution
memory is set imposed this requirement on every single job run
in the test body. This does not apply for SQL's external sort,
however, because one of the jobs does a sample and so does not
update peak execution memory.

(3) We were getting accumulators from executors that were not
registered on the driver. Not exactly sure what the cause is but
it could very well have to do with GC on the driver since we use
weak references there. We shouldn't crash the scheduler if this
happens.

Commit:
Remove unused hostname from TaskMetrics

Commit:
Reinitialize OutputMetrics et al during reconstruction

Such that downstream listeners can access their values.
This commit also generalizes the internal accumulator type from
Long to anything, since we need to store the read and write
methods of InputMetrics and OutputMetrics respectively.

Commit:
Fix *ShuffleSuite

This fixes a bug where when we reconstruct TaskMetrics we just
pass in mutable accumulators, such that when new tasks come in
they change the values of the old tasks. A more subtle bug here
is that we were passing in the accumulated values instead of the
local task values. Both are now fixed.

TODO: write a test for all of these please.

Commit:
Fix DAGSchedulerSuite

The fake accumulator values should no longer all be Longs. Ugh.

Commit:
Simplify accumulator update code a little

Commit:
Track updated blocks as an accumulator

This should be the last thing left to port everything in
TaskMetrics over to accumulators.

TODO: add some tests.

Commit:
Minor clean ups + docs

Commit:
Replace all accumulator updates with Seq[AccumulableInfo]

This makes it more consistent everywhere and simplifies some code.
E.g. the code where we set the result size is now moved back to
TaskResultGetter because of the decrease in code dependency.

This also allows us to rewrite the heartbeat events using
AccumulableInfo so we don't have to send TaskMetrics directly.

Commit:
Always send Seq[AccumulableInfo] instead of TaskMetrics

Now heartbeats go through the former. This is important because
accumulators use how many times they are serialized to determine
whether they're on the driver or the executor.

A lot of code cleanup opportunities here...

Commit:
Fix JsonProtocolSuite + cleanup

Commit:
More cleanups

Commit:
Merge branch 'master' of github.com:apache/spark into task-metrics-to-accums

SparkQA commented:
**[Test build #49207 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49207/consoleFull)** for PR 10717 at commit [`61fc9e1`](https://github.com/apache/spark/commit/61fc9e12bb46983a086f63c126c70120841631ab).

SparkQA commented:
**[Test build #49207 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49207/consoleFull)** for PR 10717 at commit [`61fc9e1`](https://github.com/apache/spark/commit/61fc9e12bb46983a086f63c126c70120841631ab).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49207/
Test FAILed.

Commit:
Fix style

SparkQA commented:
**[Test build #49225 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49225/consoleFull)** for PR 10717 at commit [`4d681fa`](https://github.com/apache/spark/commit/4d681fa04f2c4d8ea5d22b03a15724e7394cbc0e).

SparkQA commented:
**[Test build #49225 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49225/consoleFull)** for PR 10717 at commit [`4d681fa`](https://github.com/apache/spark/commit/4d681fa04f2c4d8ea5d22b03a15724e7394cbc0e).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49225/
Test FAILed.

Commit:
Fix MiMa

Commit:
Merge branch 'master' of github.com:apache/spark into task-metrics-to-accums

Conflicts:
	project/MimaExcludes.scala

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49250/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49251/
Test FAILed.

andrewor14 commented:
retest this please

SparkQA commented:
**[Test build #49255 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49255/consoleFull)** for PR 10717 at commit [`67153e6`](https://github.com/apache/spark/commit/67153e68cb455c5e99bd0374a9a6d828f0e53c66).

SparkQA commented:
**[Test build #49255 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49255/consoleFull)** for PR 10717 at commit [`67153e6`](https://github.com/apache/spark/commit/67153e68cb455c5e99bd0374a9a6d828f0e53c66).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class PageView(val url: String, val status: Int, val zipCode: Int, val userID: Int)`
  * `class FlumeEventServer(receiver: FlumeReceiver) extends AvroSourceProtocol `
  * `class UnsafeCartesianRDD(left: RDD[UnsafeRow], right: RDD[UnsafeRow], numFieldsOfRight: Int)`
  * `case class JdbcType(databaseTypeDefinition: String, jdbcNullType: Int)`
  * `class ObjectInputStreamWithLoader(_inputStream: InputStream, loader: ClassLoader)`
  * `class ConstantInputDStream[T: ClassTag](_ssc: StreamingContext, rdd: RDD[T])`
  * `abstract class InputDStream[T: ClassTag] (_ssc: StreamingContext)`
  * `abstract class ReceiverInputDStream[T: ClassTag](_ssc: StreamingContext)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49255/
Test FAILed.

Commit:
Fix MiMa

Commit:
Rename updatedBlocks -> updatedBlockStatuses

Just for consistency.

SparkQA commented:
**[Test build #49261 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49261/consoleFull)** for PR 10717 at commit [`c94420b`](https://github.com/apache/spark/commit/c94420b3552fd15fe786ecc02e6f8d70c8d2f316).

Commit:
Fix wrong comment by removing it :)

Commit:
Fix StringAccumulatorParam semantics

InputMetrics#readMethod is a string. We don't want to accumulate
the read method of one task with that of another task by
concatenating them, otherwise the accumulated value might look
something like "HadoopHadoopHadoopHadoopHadoop...", which makes
no sense. For our purposes, the String param now always overwrites
the old value with the new value.

SparkQA commented:
**[Test build #49267 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49267/consoleFull)** for PR 10717 at commit [`b271205`](https://github.com/apache/spark/commit/b271205c6c2b4170d5abd2ebb52f48fa8463112d).

Commit:
Relax scheduler dependency order + simplify some code

The previous dependency order between TaskMetrics and accumulators
in DAGScheduler is unnecessary. Removing this ordering constraint
simplifies the code a little and makes the diff a little smaller.

SparkQA commented:
**[Test build #49270 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49270/consoleFull)** for PR 10717 at commit [`8065b25`](https://github.com/apache/spark/commit/8065b250fffed9566fc93286a43e5e4a47dbb948).

SparkQA commented:
**[Test build #49261 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49261/consoleFull)** for PR 10717 at commit [`c94420b`](https://github.com/apache/spark/commit/c94420b3552fd15fe786ecc02e6f8d70c8d2f316).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49261/
Test FAILed.

SparkQA commented:
**[Test build #49267 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49267/consoleFull)** for PR 10717 at commit [`b271205`](https://github.com/apache/spark/commit/b271205c6c2b4170d5abd2ebb52f48fa8463112d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49267/
Test FAILed.

SparkQA commented:
**[Test build #49270 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49270/consoleFull)** for PR 10717 at commit [`8065b25`](https://github.com/apache/spark/commit/8065b250fffed9566fc93286a43e5e4a47dbb948).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49270/
Test FAILed.

andrewor14 committed:
Fix HistoryServerSuite

The problem was that the executor run time from a failed task was
not included in the one aggregated across the stage. This is
because the old test log did not have accumulator updates, but
only TaskMetrics, in the ExceptionFailure class.

It would seem that an easy fix would be to fallback to parsing
TaskMetrics and get accumulator updates from there. However, this
is insufficient because we could not reconstruct TaskMetrics
without having previously registered them on the driver.

This commit adds the functionality to reconstruct TaskMetrics
without any prior state. This implicitly requires Spark to be
able to reconstruct the Accumulator object from AccumulableInfo.
This was difficult because this would mean that we need to
keep track of the accumulator value type, which was not previously
encoded anywhere.

After this commit, we now use the correct type to de/serialize
the values of AccumulableInfo's that represent the internal
accumulators used by TaskMetrics. The two main benefits are:

- ExceptionFailure no longer breaks compatibility in event logs
- AccumulableInfo is type-safe in event logs for internal accums

Commit:
Merge github.com:apache/spark into task-metrics-to-accums

Commit:
Fix TaskMetricsSuite

Do not create ShuffleReadMetrics when there is no shuffle read.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49367/
Test FAILed.

SparkQA commented:
**[Test build #49371 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49371/consoleFull)** for PR 10717 at commit [`8b99071`](https://github.com/apache/spark/commit/8b99071f1591abce3cb61651413747188fb47ec5).

Commit:
Fix JsonProtocolSuite

We were asserting the JSON serialized forms of the events, which
may have out-of-order accumulators in StageInfo. Instead we should
make it a SparkListenerEvent again and assert equality there.

SparkQA commented:
**[Test build #49372 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49372/consoleFull)** for PR 10717 at commit [`dd6d8ab`](https://github.com/apache/spark/commit/dd6d8ab72b3ddb025728365bb1e8116a7da67199).

SparkQA commented:
**[Test build #49371 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49371/consoleFull)** for PR 10717 at commit [`8b99071`](https://github.com/apache/spark/commit/8b99071f1591abce3cb61651413747188fb47ec5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49371/
Test FAILed.

SparkQA commented:
**[Test build #49372 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49372/consoleFull)** for PR 10717 at commit [`dd6d8ab`](https://github.com/apache/spark/commit/dd6d8ab72b3ddb025728365bb1e8116a7da67199).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49372/
Test FAILed.

Commit:
Merge branch 'master' of github.com:apache/spark into task-metrics-to-accums

Conflicts:
	core/src/main/scala/org/apache/spark/TaskContextImpl.scala
	core/src/main/scala/org/apache/spark/scheduler/Task.scala
	core/src/test/scala/org/apache/spark/CacheManagerSuite.scala

Commit:
Fix TaskContextSuite + add some param docs

Commit:
Fix StagePageSuite

Commit:
Fix ExternalSorterSuite

We never passed "countFailedValues" to AccumulableInfo.

Commit:
Fix SQLMetricsSuite

When we reconstruct the TaskMetrics we would copy the Accumulable
without preserving the ID. However, the SQLListener downstream
expects these internal metric accumulators to use the same IDs.

The fix is to preserve the ID of the Accumulable when doing the
copying.

JoshRosen commented on project/MimaExcludes.scala:
```diff
@@ -147,6 +147,16 @@ object MimaExcludes {
+          "org.apache.spark.TaskContext.collectInternalAccumulators"),
+        ProblemFilters.exclude[MissingMethodProblem](
+          "org.apache.spark.TaskContext.collectAccumulators"),
+        ProblemFilters.exclude[MissingMethodProblem]("org.apache.spark.Accumulable.this"),
```
Do you remember which change triggered this error?

Commit:
Fix PartitionBatchPruningSuite

This test suite uses InMemoryColumnarTableScan, which internally
creates some accumulators. There were two problems:

(1) These accumulators were created before all tests, but after
each test we clear them from the global map. Now we create these
accumulators before *each* test instead.

(2) These accumulators were not updated on the driver because
we overwrite the registered accumulator every time we do a copy,
e.g. when we reconstruct TaskMetrics from accumulator updates.
We do a check in register now to avoid overwriting the original
accumulator.

Commit:
Clean up Accumulators object a little + some docs

Commit:
Fix DAGSchedulerSuite, trivial omission

SparkQA commented:
**[Test build #49424 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49424/consoleFull)** for PR 10717 at commit [`b31862a`](https://github.com/apache/spark/commit/b31862a089b20e516c77b558917c4210cdb31cfd).

SparkQA commented:
**[Test build #49424 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49424/consoleFull)** for PR 10717 at commit [`b31862a`](https://github.com/apache/spark/commit/b31862a089b20e516c77b558917c4210cdb31cfd).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49424/
Test FAILed.

Commit:
Fix style + default

SparkQA commented:
**[Test build #49430 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49430/consoleFull)** for PR 10717 at commit [`3dcc4e1`](https://github.com/apache/spark/commit/3dcc4e16acd0c2ceaa264ad23abcf067935d6a8f).

Commit:
Split Accumulators.scala into 3 files

... mainly for readability.

SparkQA commented:
**[Test build #49430 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49430/consoleFull)** for PR 10717 at commit [`3dcc4e1`](https://github.com/apache/spark/commit/3dcc4e16acd0c2ceaa264ad23abcf067935d6a8f).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49430/
Test FAILed.

Commit:
Fix MiMa

SparkQA commented:
**[Test build #49431 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49431/consoleFull)** for PR 10717 at commit [`fe451e5`](https://github.com/apache/spark/commit/fe451e5d516413347ed7bd7ebd94eb25088de60e).

Commit:
Minor cleanups

SparkQA commented:
**[Test build #49431 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49431/consoleFull)** for PR 10717 at commit [`fe451e5`](https://github.com/apache/spark/commit/fe451e5d516413347ed7bd7ebd94eb25088de60e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49431/
Test FAILed.

Commit:
Fix style

SparkQA commented:
**[Test build #49435 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49435/consoleFull)** for PR 10717 at commit [`5ca949f`](https://github.com/apache/spark/commit/5ca949f23170ca8b1763800ca106024ff29c85cf).

Commit:
Add InternalAccumulatorsSuite

SparkQA commented:
**[Test build #49435 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49435/consoleFull)** for PR 10717 at commit [`5ca949f`](https://github.com/apache/spark/commit/5ca949f23170ca8b1763800ca106024ff29c85cf).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49435/
Test FAILed.

Commit:
Fix JsonProtocolSuite

Commit:
Merge branch 'master' of github.com:apache/spark into task-metrics-to-accums

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49447/
Test FAILed.

andrewor14 commented:
retest this please

SparkQA commented:
**[Test build #49469 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49469/consoleFull)** for PR 10717 at commit [`ce10ea1`](https://github.com/apache/spark/commit/ce10ea169cb37611f0e7e36c5303f3ab454fcc7a).

SparkQA commented:
**[Test build #49469 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49469/consoleFull)** for PR 10717 at commit [`ce10ea1`](https://github.com/apache/spark/commit/ce10ea169cb37611f0e7e36c5303f3ab454fcc7a).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait ObjectOperator extends LogicalPlan `
  * `case class MapPartitions(`
  * `case class AppendColumns(`
  * `case class MapGroups(`
  * `case class CoGroup(`
  * `trait ObjectOperator extends SparkPlan `
  * `case class MapPartitions(`
  * `case class AppendColumns(`
  * `case class MapGroups(`
  * `case class CoGroup(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49469/
Test FAILed.

Commit:
Add more tests to AccumulatorSuite

Commit:
Fix style

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49474/
Test FAILed.

Commit:
Add test for accum updates during failures

Commit:
Merge branch 'master' of github.com:apache/spark into task-metrics-to-accums

Conflicts:
	core/src/test/scala/org/apache/spark/storage/StorageStatusListenerSuite.scala
	core/src/test/scala/org/apache/spark/ui/storage/StorageTabSuite.scala
	core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala

andrewor14 commented:
retest this please

SparkQA commented:
**[Test build #49482 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49482/consoleFull)** for PR 10717 at commit [`23af334`](https://github.com/apache/spark/commit/23af334f2ab7c7fbd54ff9ca2b0bc547f5187937).

SparkQA commented:
**[Test build #49480 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49480/consoleFull)** for PR 10717 at commit [`23af334`](https://github.com/apache/spark/commit/23af334f2ab7c7fbd54ff9ca2b0bc547f5187937).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49482/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49480/
Test FAILed.

andrewor14 commented:
retest this please

SparkQA commented:
**[Test build #49485 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49485/consoleFull)** for PR 10717 at commit [`b58f2e6`](https://github.com/apache/spark/commit/b58f2e64e05e3e7f89ff8da4823007d810fc98b4).

nongli commented:
This patch is really difficult to review since it combines simple refactoring and more subtle logic. Is it possible to break this up? If there were patches that did simple clean up, those can get reviewed and merged very quickly.

andrewor14 commented:
Yeah it's still WIP but once it's in a more ready state I'll try to break it up.

Commit:
Add tests for TaskMetrics

SparkQA commented:
**[Test build #49485 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49485/consoleFull)** for PR 10717 at commit [`b58f2e6`](https://github.com/apache/spark/commit/b58f2e64e05e3e7f89ff8da4823007d810fc98b4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49485/
Test FAILed.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49504/
Test FAILed.

Commit:
Add fine-grained test for collecting accumulators during failures

SparkQA commented:
**[Test build #49509 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49509/consoleFull)** for PR 10717 at commit [`00a12a4`](https://github.com/apache/spark/commit/00a12a468e1f83f6fce61b88ce494059e512df21).

SparkQA commented:
**[Test build #49509 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49509/consoleFull)** for PR 10717 at commit [`00a12a4`](https://github.com/apache/spark/commit/00a12a468e1f83f6fce61b88ce494059e512df21).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49509/
Test FAILed.

Commit:
Fix style

SparkQA commented:
**[Test build #49601 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49601/consoleFull)** for PR 10717 at commit [`ed9de9c`](https://github.com/apache/spark/commit/ed9de9c084b0a8df41b24768404c63cb331c8a25).

SparkQA commented:
**[Test build #49601 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49601/consoleFull)** for PR 10717 at commit [`ed9de9c`](https://github.com/apache/spark/commit/ed9de9c084b0a8df41b24768404c63cb331c8a25).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49601/
Test FAILed.

andrewor14 commented:
OK, I broke this down into smaller PRs so far:

#10810 - move classes to their own files for readability
#10811 - rename 3 fields in ShuffleWriteMetrics
#10815 - refactor TaskMetrics to make it not use public var's

More coming soon, I'll keep this comment updated.

andrewor14 commented:
By the end of all the smaller patches there will be too much conflict to resolve, so I'm closing this.

andrewor14 commented:
Note to self: DO NOT DELETE THIS BRANCH!!
